File: deno/packages/agents/seal/knowledge_integration.ts
/**
 * SEAL + Knowledge System integration.
 *
 * Bridges SEAL's entity-centric learning with a Knowledge System's behavioral
 * truths (BKS) and personal facts (PKS). On the Rust side this module talks to
 * concrete `BehavioralKnowledgeCache` / `PersonalKnowledgeCache` types backed
 * by LanceDB. On the Deno side the `@brainwires/knowledge` package is
 * interface-only — the concrete BKS/PKS work happens Rust-side — so this
 * module is structured as interface wiring: callers inject the cache shapes
 * (typically via an RPC bridge) and the coordinator glues them to SEAL.
 *
 * Equivalent to Rust's `brainwires_agents::seal::knowledge_integration` module.
 */

import type { QueryPattern } from "./learning.ts";
import type { LearningCoordinator, PatternHint } from "./learning.ts";
import type { QuestionType } from "./query_core.ts";
import type { ResolvedReference, SealProcessingResult } from "./types.ts";

// ─── Types expected from @brainwires/knowledge (interfaces) ────────────────

/** Category of a behavioral truth. */
export type TruthCategory =
  | "command_usage"
  | "task_strategy"
  | "error_recovery"
  | "convention"
  | "capability"
  | "other";

/** How a behavioral truth was sourced. */
export type TruthSource =
  | "manual"
  | "success_pattern"
  | "failure_pattern"
  | "user_correction";

/** A behavioral truth entry (shape matches Rust `BehavioralTruth`). */
export interface BehavioralTruth {
  category: TruthCategory;
  context_pattern: string;
  rule: string;
  rationale: string;
  source: TruthSource;
  confidence: number;
  author: string | undefined;
}

export function newBehavioralTruth(
  category: TruthCategory,
  context_pattern: string,
  rule: string,
  rationale: string,
  source: TruthSource,
  author: string | undefined,
): BehavioralTruth {
  return {
    category,
    context_pattern,
    rule,
    rationale,
    source,
    confidence: 1.0,
    author,
  };
}

/** A scored (truth, relevance) pair from a BKS lookup. */
export type ScoredTruth = [BehavioralTruth, number];

/** A personal fact entry. */
export interface PersonalFact {
  key: string;
  value: string;
  confidence: number;
  deleted: boolean;
}

/**
 * Behavioral knowledge cache — concrete implementation lives Rust-side.
 * Deno consumers inject an object matching this shape (e.g., an RPC bridge).
 */
export interface BehavioralKnowledgeCache {
  getMatchingTruthsWithScores(
    query: string,
    min_confidence: number,
    limit: number,
  ): Promise<ScoredTruth[]>;
  getReliableTruths(
    min_confidence: number,
    days: number,
  ): Promise<BehavioralTruth[]>;
  queueSubmission(truth: BehavioralTruth): Promise<void>;
}

/** Personal knowledge cache — concrete implementation lives Rust-side. */
export interface PersonalKnowledgeCache {
  getAllFacts(): Promise<PersonalFact[]>;
  upsertFactSimple(
    key: string,
    value: string,
    confidence: number,
    local_only: boolean,
  ): Promise<void>;
}

// ─── Config ─────────────────────────────────────────────────────────────────

export type EntityResolutionStrategy =
  | { kind: "seal_first" }
  | { kind: "pks_context_first" }
  | { kind: "hybrid"; seal_weight: number; pks_weight: number };

/** Configuration for SEAL + Knowledge integration. */
export interface IntegrationConfig {
  enabled: boolean;
  seal_to_knowledge: boolean;
  knowledge_to_seal: boolean;
  min_seal_quality_for_bks_boost: number;
  min_seal_quality_for_pks_boost: number;
  pattern_promotion_threshold: number;
  min_pattern_uses: number;
  cache_bks_in_seal: boolean;
  entity_resolution_strategy: EntityResolutionStrategy;
  seal_weight: number;
  bks_weight: number;
  pks_weight: number;
}

export const DEFAULT_PATTERN_PROMOTION_THRESHOLD = 0.8;

export function defaultIntegrationConfig(): IntegrationConfig {
  return {
    enabled: true,
    seal_to_knowledge: true,
    knowledge_to_seal: true,
    min_seal_quality_for_bks_boost: 0.7,
    min_seal_quality_for_pks_boost: 0.5,
    pattern_promotion_threshold: DEFAULT_PATTERN_PROMOTION_THRESHOLD,
    min_pattern_uses: 5,
    cache_bks_in_seal: true,
    entity_resolution_strategy: {
      kind: "hybrid",
      seal_weight: 0.6,
      pks_weight: 0.4,
    },
    seal_weight: 0.5,
    bks_weight: 0.3,
    pks_weight: 0.2,
  };
}

export function integrationConfigSealToKnowledgeOnly(): IntegrationConfig {
  return { ...defaultIntegrationConfig(), knowledge_to_seal: false };
}

export function integrationConfigDisabled(): IntegrationConfig {
  return { ...defaultIntegrationConfig(), enabled: false };
}

export function validateIntegrationConfig(c: IntegrationConfig): void {
  if (c.min_seal_quality_for_bks_boost < 0 || c.min_seal_quality_for_bks_boost > 1) {
    throw new Error("min_seal_quality_for_bks_boost must be between 0.0 and 1.0");
  }
  if (c.pattern_promotion_threshold < 0 || c.pattern_promotion_threshold > 1) {
    throw new Error("pattern_promotion_threshold must be between 0.0 and 1.0");
  }
  const sum = c.seal_weight + c.bks_weight + c.pks_weight;
  if (Math.abs(sum - 1.0) > 0.01) {
    throw new Error(
      `Confidence weights must sum to 1.0 (got: ${sum.toFixed(2)})`,
    );
  }
}

// ─── Coordinator ────────────────────────────────────────────────────────────

/**
 * Coordinator bridging SEAL with a knowledge system. Deno callers supply
 * cache implementations — usually thin proxies to a Rust-side BKS/PKS server.
 */
export class SealKnowledgeCoordinator {
  private bks_cache: BehavioralKnowledgeCache;
  private pks_cache: PersonalKnowledgeCache;
  readonly integrationConfig: IntegrationConfig;

  constructor(
    bks_cache: BehavioralKnowledgeCache,
    pks_cache: PersonalKnowledgeCache,
    config: IntegrationConfig = defaultIntegrationConfig(),
  ) {
    validateIntegrationConfig(config);
    this.bks_cache = bks_cache;
    this.pks_cache = pks_cache;
    this.integrationConfig = config;
  }

  /** PKS context for SEAL entity resolutions. */
  async getPksContext(
    seal_result: SealProcessingResult,
  ): Promise<string | undefined> {
    if (!this.integrationConfig.enabled) return undefined;
    if (
      seal_result.quality_score <
        this.integrationConfig.min_seal_quality_for_pks_boost
    ) return undefined;

    const entities = seal_result.resolutions.map((r) => r.antecedent);
    if (entities.length === 0) return undefined;

    const allFacts = await this.pks_cache.getAllFacts();
    const contextParts: string[] = [];

    for (const entity of entities) {
      const relevant = allFacts.filter((f) =>
        !f.deleted && (f.key.includes(entity) || f.value.includes(entity))
      );
      if (relevant.length === 0) continue;
      contextParts.push(`\n**${entity}:**`);
      for (const fact of relevant) {
        if (fact.confidence >= 0.5) {
          contextParts.push(
            `  - ${fact.value} (confidence: ${fact.confidence.toFixed(2)})`,
          );
        }
      }
    }

    if (contextParts.length === 0) return undefined;
    return `# PERSONAL CONTEXT\n\nRelevant facts about entities mentioned:\n${
      contextParts.join("\n")
    }`;
  }

  /** BKS context for query. */
  async getBksContext(query: string): Promise<string | undefined> {
    if (!this.integrationConfig.enabled) return undefined;
    const truths = await this.bks_cache.getMatchingTruthsWithScores(query, 0.5, 5);
    if (truths.length === 0) return undefined;

    const parts: string[] = ["# BEHAVIORAL KNOWLEDGE\n"];
    parts.push("Learned patterns that may be relevant:\n");
    for (const [truth, score] of truths) {
      parts.push(
        `\n**${truth.context_pattern}** (confidence: ${
          truth.confidence.toFixed(2)
        }, relevance: ${score.toFixed(2)}):`,
      );
      parts.push(`  Rule: ${truth.rule}`);
      parts.push(`  Why: ${truth.rationale}`);
    }
    return parts.join("\n");
  }

  /** Combine SEAL quality with optional BKS and PKS confidences. */
  harmonizeConfidence(
    seal_quality: number,
    bks_confidence: number | undefined,
    pks_confidence: number | undefined,
  ): number {
    let weightedSum = seal_quality * this.integrationConfig.seal_weight;
    let totalWeight = this.integrationConfig.seal_weight;
    if (bks_confidence !== undefined) {
      weightedSum += bks_confidence * this.integrationConfig.bks_weight;
      totalWeight += this.integrationConfig.bks_weight;
    }
    if (pks_confidence !== undefined) {
      weightedSum += pks_confidence * this.integrationConfig.pks_weight;
      totalWeight += this.integrationConfig.pks_weight;
    }
    return totalWeight > 0 ? Math.min(weightedSum / totalWeight, 1.0) : seal_quality;
  }

  /** Quality-adjusted retrieval threshold. */
  adjustRetrievalThreshold(base: number, seal_quality: number): number {
    return base * Math.max(0.7 + 0.3 * seal_quality, 0.5);
  }

  /** Check promotion criteria and, if met, queue a BKS submission. */
  async checkAndPromotePattern(
    pattern: QueryPattern,
    execution_context: string,
  ): Promise<BehavioralTruth | undefined> {
    if (
      !this.integrationConfig.enabled ||
      !this.integrationConfig.seal_to_knowledge
    ) return undefined;
    if (pattern.reliability() < this.integrationConfig.pattern_promotion_threshold) {
      return undefined;
    }
    const total = pattern.success_count + pattern.failure_count;
    if (total < this.integrationConfig.min_pattern_uses) return undefined;

    const truth = newBehavioralTruth(
      this.inferCategory(pattern.question_type),
      execution_context,
      this.generalizePatternToRule(pattern),
      `Learned from ${pattern.success_count} successful executions with ${
        (pattern.reliability() * 100).toFixed(1)
      }% reliability (SEAL pattern)`,
      "success_pattern",
      undefined,
    );
    await this.bks_cache.queueSubmission(truth);
    return truth;
  }

  /** Sync high-confidence BKS truths into SEAL's global memory. */
  async syncBksToSeal(seal_learning: LearningCoordinator): Promise<number> {
    if (
      !this.integrationConfig.enabled ||
      !this.integrationConfig.knowledge_to_seal ||
      !this.integrationConfig.cache_bks_in_seal
    ) return 0;

    const truths = await this.bks_cache.getReliableTruths(0.7, 30);
    let loaded = 0;
    for (const truth of truths) {
      const hint = this.truthToPatternHint(truth);
      if (hint !== undefined) {
        seal_learning.global.addPatternHint(hint);
        loaded += 1;
      }
    }
    return loaded;
  }

  /** Track SEAL entity resolutions as recent-entity facts in PKS. */
  async observeSealResolutions(
    resolutions: readonly ResolvedReference[],
  ): Promise<void> {
    if (!this.integrationConfig.enabled) return;
    for (const resolution of resolutions) {
      const key = `recent_entity:${resolution.antecedent}`;
      await this.pks_cache.upsertFactSimple(
        key,
        resolution.antecedent,
        resolution.confidence,
        true,
      );
    }
  }

  /** Record a tool failure pattern as a BKS entry. */
  async recordToolFailure(
    tool_name: string,
    error_message: string,
    context: string,
  ): Promise<void> {
    if (
      !this.integrationConfig.enabled ||
      !this.integrationConfig.seal_to_knowledge
    ) return;
    const truth = newBehavioralTruth(
      "error_recovery",
      context,
      `Tool '${tool_name}' commonly fails with: ${error_message}`,
      "Observed from validation failures",
      "failure_pattern",
      undefined,
    );
    await this.bks_cache.queueSubmission(truth);
  }

  /** Accessors mirroring Rust `get_pks_cache` / `get_bks_cache`. */
  getPksCache(): PersonalKnowledgeCache {
    return this.pks_cache;
  }
  getBksCache(): BehavioralKnowledgeCache {
    return this.bks_cache;
  }

  private inferCategory(qt: QuestionType): TruthCategory {
    switch (qt) {
      case "definition":
        return "command_usage";
      case "dependency":
      case "location":
      case "count":
      case "superlative":
      default:
        return "task_strategy";
    }
  }

  private generalizePatternToRule(pattern: QueryPattern): string {
    const typesStr = pattern.required_types.join(", ");
    return `For '${pattern.question_type}' queries about ${typesStr}, use pattern: ${pattern.template}`;
  }

  private truthToPatternHint(t: BehavioralTruth): PatternHint | undefined {
    return {
      context_pattern: t.context_pattern,
      rule: t.rule,
      confidence: t.confidence,
      source: "bks",
    };
  }
}

File: crates/brainwires-seal/src/knowledge_integration.rs
//! SEAL + Knowledge System Integration
//!
//! This module bridges SEAL's entity-centric learning with the Knowledge System's
//! behavioral truths (BKS) and personal facts (PKS), enabling bidirectional learning
//! and context-aware entity resolution.
//!
//! ## Key Integration Points
//!
//! 1. **SEAL → PKS**: Entity resolutions trigger lookups of personal facts
//! 2. **SEAL → BKS**: Query context triggers lookups of behavioral truths
//! 3. **SEAL → BKS Promotion**: High-reliability patterns promoted to shared knowledge
//! 4. **BKS → SEAL**: Behavioral truths loaded into SEAL's global memory
//! 5. **Quality-Aware Context**: SEAL quality scores adjust retrieval thresholds
//!
//! ## Architecture
//!
//! ```text
//! User Input
//!     │
//!     ▼
//! ┌────────────────────────────────────────────┐
//! │   SealKnowledgeCoordinator                 │
//! │                                            │
//! │   ┌─────────────────────────────────────┐ │
//! │   │ SEAL Preprocessing                   │ │
//! │   │ • Coreference: "it" → "main.rs"     │ │
//! │   │ • Query extraction: S-expressions   │ │
//! │   │ • Quality score: 0.0-1.0            │ │
//! │   └──────────┬──────────────────────────┘ │
//! │              │                             │
//! │              ▼                             │
//! │   ┌─────────────────────────────────────┐ │
//! │   │ Knowledge Lookup (PARALLEL)         │ │
//! │   │ • PKS: entity facts (main.rs)       │ │
//! │   │ • BKS: behavioral truths (rust)     │ │
//! │   └──────────┬──────────────────────────┘ │
//! │              │                             │
//! │              ▼                             │
//! │   ┌─────────────────────────────────────┐ │
//! │   │ Confidence Harmonization            │ │
//! │   │ • Combine: SEAL + BKS + PKS         │ │
//! │   │ • Adjust thresholds by quality      │ │
//! │   └──────────┬──────────────────────────┘ │
//! └──────────────┼──────────────────────────────┘
//!                │
//!                ▼
//!     Enhanced Context → OrchestratorAgent
//! ```

use super::{QueryPattern, ResolvedReference, SealProcessingResult};
use anyhow::Result;
use brainwires_knowledge::knowledge::bks_pks::{
    BehavioralKnowledgeCache, BehavioralTruth, PersonalKnowledgeCache, TruthCategory, TruthSource,
};
use std::sync::Arc;
use tokio::sync::Mutex;

/// Default minimum pattern reliability for BKS promotion (80% success rate).
const DEFAULT_PATTERN_PROMOTION_THRESHOLD: f32 = 0.8;

/// Configuration for SEAL + Knowledge integration
#[derive(Debug, Clone)]
pub struct IntegrationConfig {
    /// Master toggle for the integration
    pub enabled: bool,

    /// Enable SEAL patterns → BKS promotion
    pub seal_to_knowledge: bool,

    /// Enable BKS truths → SEAL pattern loading
    pub knowledge_to_seal: bool,

    /// Minimum SEAL quality score to boost with BKS context (0.0-1.0)
    /// Default: 0.7 (only high-quality SEAL results get BKS boost)
    pub min_seal_quality_for_bks_boost: f32,

    /// Minimum SEAL quality score to boost with PKS context (0.0-1.0)
    /// Default: 0.5 (medium-quality SEAL results get PKS boost)
    pub min_seal_quality_for_pks_boost: f32,

    /// Minimum pattern reliability for BKS promotion (0.0-1.0)
    /// Default: 0.8 (80% success rate required)
    pub pattern_promotion_threshold: f32,

    /// Minimum pattern uses before considering promotion
    /// Default: 5 (need statistical significance)
    pub min_pattern_uses: u32,

    /// Cache BKS truths in SEAL's global memory
    pub cache_bks_in_seal: bool,

    /// Entity resolution strategy
    pub entity_resolution_strategy: EntityResolutionStrategy,

    /// Weight for SEAL quality in confidence harmonization (0.0-1.0)
    /// Default: 0.5 (50% weight)
    pub seal_weight: f32,

    /// Weight for BKS confidence in harmonization (0.0-1.0)
    /// Default: 0.3 (30% weight)
    pub bks_weight: f32,

    /// Weight for PKS confidence in harmonization (0.0-1.0)
    /// Default: 0.2 (20% weight)
    pub pks_weight: f32,
}

impl Default for IntegrationConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            seal_to_knowledge: true,
            knowledge_to_seal: true,
            min_seal_quality_for_bks_boost: 0.7,
            min_seal_quality_for_pks_boost: 0.5,
            pattern_promotion_threshold: DEFAULT_PATTERN_PROMOTION_THRESHOLD,
            min_pattern_uses: 5,
            cache_bks_in_seal: true,
            entity_resolution_strategy: EntityResolutionStrategy::Hybrid {
                seal_weight: 0.6,
                pks_weight: 0.4,
            },
            seal_weight: 0.5,
            bks_weight: 0.3,
            pks_weight: 0.2,
        }
    }
}

impl IntegrationConfig {
    /// Create config with all features enabled (recommended)
    pub fn full() -> Self {
        Self::default()
    }

    /// Create config with only SEAL → Knowledge enabled (no BKS → SEAL loading)
    pub fn seal_to_knowledge_only() -> Self {
        Self {
            knowledge_to_seal: false,
            ..Self::default()
        }
    }

    /// Create config with integration disabled
    pub fn disabled() -> Self {
        Self {
            enabled: false,
            ..Self::default()
        }
    }

    /// Validate configuration
    pub fn validate(&self) -> Result<()> {
        if self.min_seal_quality_for_bks_boost < 0.0 || self.min_seal_quality_for_bks_boost > 1.0 {
            anyhow::bail!("min_seal_quality_for_bks_boost must be between 0.0 and 1.0");
        }

        if self.pattern_promotion_threshold < 0.0 || self.pattern_promotion_threshold > 1.0 {
            anyhow::bail!("pattern_promotion_threshold must be between 0.0 and 1.0");
        }

        let weight_sum = self.seal_weight + self.bks_weight + self.pks_weight;
        if (weight_sum - 1.0).abs() > 0.01 {
            anyhow::bail!(
                "Confidence weights must sum to 1.0 (got: {:.2})",
                weight_sum
            );
        }

        Ok(())
    }
}

/// Strategy for resolving entity references when both SEAL and PKS have candidates
#[derive(Debug, Clone)]
pub enum EntityResolutionStrategy {
    /// Always prefer SEAL's resolution
    SealFirst,

    /// Always prefer PKS context-based resolution
    PksContextFirst,

    /// Weighted combination of SEAL and PKS confidence
    Hybrid {
        /// Weight given to SEAL confidence scores.
        seal_weight: f32,
        /// Weight given to PKS confidence scores.
        pks_weight: f32,
    },
}

/// Bridges SEAL's entity-centric learning with Knowledge System's behavioral truths
pub struct SealKnowledgeCoordinator {
    /// BKS cache for behavioral truths
    bks_cache: Arc<Mutex<BehavioralKnowledgeCache>>,

    /// PKS cache for personal facts
    pks_cache: Arc<Mutex<PersonalKnowledgeCache>>,

    /// Integration settings
    config: IntegrationConfig,
}

impl SealKnowledgeCoordinator {
    /// Create a new coordinator
    pub fn new(
        bks_cache: Arc<Mutex<BehavioralKnowledgeCache>>,
        pks_cache: Arc<Mutex<PersonalKnowledgeCache>>,
        config: IntegrationConfig,
    ) -> Result<Self> {
        config.validate()?;

        Ok(Self {
            bks_cache,
            pks_cache,
            config,
        })
    }

    /// Create with default configuration
    pub fn with_defaults(
        bks_cache: Arc<Mutex<BehavioralKnowledgeCache>>,
        pks_cache: Arc<Mutex<PersonalKnowledgeCache>>,
    ) -> Result<Self> {
        Self::new(bks_cache, pks_cache, IntegrationConfig::default())
    }

    /// Get PKS context for SEAL entity resolutions
    ///
    /// Given SEAL's entity resolutions, look up relevant personal facts
    /// that provide context about those entities.
    ///
    /// Example:
    /// - SEAL resolves "it" → "main.rs"
    /// - PKS has: "main.rs is entry point for brainwires-cli"
    /// - Returns: Formatted context string with relevant facts
    pub async fn get_pks_context(
        &self,
        seal_result: &SealProcessingResult,
    ) -> Result<Option<String>> {
        if !self.config.enabled {
            return Ok(None);
        }

        // Check if SEAL quality is high enough for PKS boost
        if seal_result.quality_score < self.config.min_seal_quality_for_pks_boost {
            return Ok(None);
        }

        // Extract entities from SEAL resolutions
        let entities: Vec<&str> = seal_result
            .resolutions
            .iter()
            .map(|r| r.antecedent.as_str())
            .collect();

        if entities.is_empty() {
            return Ok(None);
        }

        // Look up facts for each entity
        let pks = self.pks_cache.lock().await;
        let mut context_parts = Vec::new();

        for entity in entities {
            // Get facts related to this entity by looking for keys containing the entity name
            // This is a simple heuristic - could be improved with fuzzy matching
            let all_facts: Vec<_> = pks
                .get_all_facts()
                .into_iter()
                .filter(|f| !f.deleted && (f.key.contains(entity) || f.value.contains(entity)))
                .collect();

            if !all_facts.is_empty() {
                context_parts.push(format!("\n**{}:**", entity));

                for fact in all_facts {
                    // Filter by confidence (only include reliable facts)
                    if fact.confidence >= 0.5 {
                        context_parts.push(format!(
                            "  - {} (confidence: {:.2})",
                            fact.value, fact.confidence
                        ));
                    }
                }
            }
        }

        if context_parts.is_empty() {
            Ok(None)
        } else {
            Ok(Some(format!(
                "# PERSONAL CONTEXT\n\nRelevant facts about entities mentioned:\n{}",
                context_parts.join("\n")
            )))
        }
    }

    /// Get BKS context for query
    ///
    /// Given the user's query (after SEAL processing), look up relevant
    /// behavioral truths that might help with execution.
    ///
    /// Example:
    /// - Query: "How do I run the Rust project?"
    /// - BKS has: "For Rust projects, use 'cargo run' not 'rustc main.rs'"
    /// - Returns: Formatted context string with relevant truths
    pub async fn get_bks_context(&self, query: &str) -> Result<Option<String>> {
        if !self.config.enabled {
            return Ok(None);
        }

        let bks = self.bks_cache.lock().await;

        // Get truths matching the query context with scores
        let truths = bks.get_matching_truths_with_scores(query, 0.5, 5)?;

        if truths.is_empty() {
            return Ok(None);
        }

        let mut context_parts = vec!["# BEHAVIORAL KNOWLEDGE\n".to_string()];
        context_parts.push("Learned patterns that may be relevant:\n".to_string());

        for (truth, score) in truths {
            context_parts.push(format!(
                "\n**{}** (confidence: {:.2}, relevance: {:.2}):",
                truth.context_pattern, truth.confidence, score
            ));
            context_parts.push(format!("  Rule: {}", truth.rule));
            context_parts.push(format!("  Why: {}", truth.rationale));
        }

        Ok(Some(context_parts.join("\n")))
    }

    /// Harmonize confidence from multiple sources
    ///
    /// Combines SEAL quality score with BKS and PKS confidence values
    /// using weighted averaging to produce a unified confidence score.
    pub fn harmonize_confidence(
        &self,
        seal_quality: f32,
        bks_confidence: Option<f32>,
        pks_confidence: Option<f32>,
    ) -> f32 {
        let mut weighted_sum = seal_quality * self.config.seal_weight;
        let mut total_weight = self.config.seal_weight;

        if let Some(bks) = bks_confidence {
            weighted_sum += bks * self.config.bks_weight;
            total_weight += self.config.bks_weight;
        }

        if let Some(pks) = pks_confidence {
            weighted_sum += pks * self.config.pks_weight;
            total_weight += self.config.pks_weight;
        }

        // Normalize by total weight
        if total_weight > 0.0 {
            (weighted_sum / total_weight).min(1.0)
        } else {
            seal_quality
        }
    }

    /// Adjust retrieval threshold based on SEAL quality
    ///
    /// When SEAL quality is low, we need more context to compensate,
    /// so we lower the retrieval threshold to include more historical messages.
    ///
    /// When SEAL quality is high, we can be more selective with context.
    pub fn adjust_retrieval_threshold(&self, base_threshold: f32, seal_quality: f32) -> f32 {
        // Quality adjustment factor: lower quality → lower threshold (more context)
        // Formula: adjusted = base * (0.7 + 0.3 * quality)
        // - quality = 0.0 → 70% of base threshold
        // - quality = 1.0 → 100% of base threshold
        base_threshold * (0.7 + 0.3 * seal_quality).max(0.5)
    }

    /// Check if a SEAL pattern should be promoted to BKS
    ///
    /// Patterns are promoted when they:
    /// 1. Have reliability above threshold (default: 0.8)
    /// 2. Have been used enough times (default: 5)
    /// 3. Integration is enabled
    pub async fn check_and_promote_pattern(
        &mut self,
        pattern: &QueryPattern,
        execution_context: &str,
    ) -> Result<Option<BehavioralTruth>> {
        if !self.config.enabled || !self.config.seal_to_knowledge {
            return Ok(None);
        }

        // Check promotion criteria
        if pattern.reliability() < self.config.pattern_promotion_threshold {
            tracing::debug!(
                "Pattern '{}' reliability ({:.2}) below threshold ({:.2})",
                pattern.template,
                pattern.reliability(),
                self.config.pattern_promotion_threshold
            );
            return Ok(None);
        }

        let total_uses = pattern.success_count + pattern.failure_count;
        if total_uses < self.config.min_pattern_uses {
            tracing::debug!(
                "Pattern '{}' uses ({}) below minimum ({})",
                pattern.template,
                total_uses,
                self.config.min_pattern_uses
            );
            return Ok(None);
        }

        // Create BKS truth from SEAL pattern
        let category = self.infer_category(&pattern.question_type);
        let rule = self.generalize_pattern_to_rule(pattern);
        let rationale = format!(
            "Learned from {} successful executions with {:.1}% reliability (SEAL pattern)",
            pattern.success_count,
            pattern.reliability() * 100.0
        );

        let truth = BehavioralTruth::new(
            category,
            execution_context.to_string(),
            rule,
            rationale,
            TruthSource::SuccessPattern, // Pattern emerged from successful usage
            None,                        // Anonymous
        );

        // Submit to BKS
        let mut bks = self.bks_cache.lock().await;
        match bks.queue_submission(truth.clone()) {
            Ok(_) => {
                tracing::info!(
                    "✓ Promoted SEAL pattern to BKS: '{}' (reliability: {:.2}, uses: {})",
                    pattern.template,
                    pattern.reliability(),
                    total_uses
                );
                Ok(Some(truth))
            }
            Err(e) => {
                tracing::warn!("Failed to promote SEAL pattern to BKS: {}", e);
                Err(e)
            }
        }
    }

    /// Load relevant BKS truths into SEAL's global memory on startup
    ///
    /// This enables SEAL to benefit from collective learning by having
    /// high-confidence behavioral truths available for pattern matching.
    pub async fn sync_bks_to_seal(
        &mut self,
        seal_learning: &mut super::learning::LearningCoordinator,
    ) -> Result<u32> {
        if !self.config.enabled || !self.config.knowledge_to_seal {
            return Ok(0);
        }

        if !self.config.cache_bks_in_seal {
            return Ok(0);
        }

        let bks = self.bks_cache.lock().await;

        // Get high-confidence truths (> 0.7) from last 30 days
        let truths = bks.get_reliable_truths(0.7, 30);

        let mut loaded = 0;
        for truth in truths {
            // Convert BKS truth to structured SEAL pattern hint and store it
            if let Some(hint) = self.truth_to_pattern_hint(truth) {
                seal_learning.global.add_pattern_hint(hint);
                tracing::debug!(
                    "Loaded BKS truth into SEAL: {} -> {}",
                    truth.context_pattern,
                    truth.rule
                );
                loaded += 1;
            }
        }

        tracing::info!("Loaded {} BKS truths into SEAL global memory", loaded);
        Ok(loaded)
    }

    /// Observe SEAL resolutions for PKS learning
    ///
    /// Tracks which entities the user focuses on, allowing PKS to build
    /// a profile of user interests and recently-used entities.
    pub async fn observe_seal_resolutions(
        &mut self,
        resolutions: &[ResolvedReference],
    ) -> Result<()> {
        if !self.config.enabled {
            return Ok(());
        }

        let mut pks = self.pks_cache.lock().await;

        for resolution in resolutions {
            // Track entity as context fact
            let key = format!("recent_entity:{}", resolution.antecedent);

            // Upsert fact (will update timestamp if exists)
            pks.upsert_fact_simple(
                &key,
                &resolution.antecedent,
                resolution.confidence,
                true, // local_only (don't sync entity tracking)
            )?;
        }

        Ok(())
    }

    /// Record a tool failure pattern for shared learning
    ///
    /// When validation fails or tools error, record the pattern in BKS
    /// so other users can benefit from this knowledge.
    pub async fn record_tool_failure(
        &mut self,
        tool_name: &str,
        error_message: &str,
        context: &str,
    ) -> Result<()> {
        if !self.config.enabled || !self.config.seal_to_knowledge {
            return Ok(());
        }

        // Create behavioral truth about tool failure pattern
        let truth = BehavioralTruth::new(
            TruthCategory::ErrorRecovery,
            context.to_string(),
            format!(
                "Tool '{}' commonly fails with: {}",
                tool_name, error_message
            ),
            "Observed from validation failures".to_string(),
            TruthSource::FailurePattern,
            None,
        );

        let mut bks = self.bks_cache.lock().await;
        bks.queue_submission(truth)?;

        tracing::debug!(
            "Recorded tool failure pattern: {} in context: {}",
            tool_name,
            context
        );

        Ok(())
    }

    /// Get the current configuration
    pub fn config(&self) -> &IntegrationConfig {
        &self.config
    }

    /// Get reference to PKS cache
    pub fn get_pks_cache(&self) -> Arc<Mutex<PersonalKnowledgeCache>> {
        Arc::clone(&self.pks_cache)
    }

    /// Get reference to BKS cache
    pub fn get_bks_cache(&self) -> Arc<Mutex<BehavioralKnowledgeCache>> {
        Arc::clone(&self.bks_cache)
    }

    // --- Private helper methods ---

    /// Infer BKS category from SEAL question type
    fn infer_category(&self, question_type: &super::query_core::QuestionType) -> TruthCategory {
        use super::query_core::QuestionType;

        match question_type {
            QuestionType::Definition => TruthCategory::CommandUsage,
            QuestionType::Dependency => TruthCategory::TaskStrategy,
            QuestionType::Location => TruthCategory::TaskStrategy,
            QuestionType::Count | QuestionType::Superlative => TruthCategory::TaskStrategy,
            _ => TruthCategory::TaskStrategy,
        }
    }

    /// Generalize a SEAL pattern template into a BKS rule
    fn generalize_pattern_to_rule(&self, pattern: &QueryPattern) -> String {
        // Note: QueryPattern has required_types, not entity_types
        let types_str = pattern
            .required_types
            .iter()
            .map(|t| format!("{:?}", t))
            .collect::<Vec<_>>()
            .join(", ");

        format!(
            "For '{:?}' queries about {}, use pattern: {}",
            pattern.question_type, types_str, pattern.template
        )
    }

    /// Convert BKS truth to a structured SEAL pattern hint.
    ///
    /// This is a best-effort conversion since BKS and SEAL use different schemas.
    /// Returns None if the truth doesn't map to a SEAL pattern.
    fn truth_to_pattern_hint(
        &self,
        truth: &BehavioralTruth,
    ) -> Option<super::learning::PatternHint> {
        Some(super::learning::PatternHint {
            context_pattern: truth.context_pattern.clone(),
            rule: truth.rule.clone(),
            confidence: truth.confidence as f64,
            source: "bks".to_string(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_integration_config_validation() {
        let mut config = IntegrationConfig::default();
        assert!(config.validate().is_ok());

        // Invalid quality threshold
        config.min_seal_quality_for_bks_boost = 1.5;
        assert!(config.validate().is_err());

        // Invalid weight sum
        config = IntegrationConfig::default();
        config.seal_weight = 0.5;
        config.bks_weight = 0.5;
        config.pks_weight = 0.5; // Sum > 1.0
        assert!(config.validate().is_err());
    }

    #[test]
    fn test_confidence_harmonization() {
        let coordinator = create_test_coordinator();

        // Only SEAL
        let conf = coordinator.harmonize_confidence(0.8, None, None);
        assert!((conf - 0.8).abs() < 0.01);

        // SEAL + BKS + PKS
        let conf = coordinator.harmonize_confidence(0.6, Some(0.9), Some(0.8));
        // Expected: 0.6*0.5 + 0.9*0.3 + 0.8*0.2 = 0.3 + 0.27 + 0.16 = 0.73
        assert!((conf - 0.73).abs() < 0.01);
    }

    #[test]
    fn test_retrieval_threshold_adjustment() {
        let coordinator = create_test_coordinator();

        // Low quality → lower threshold (need more context)
        let adjusted = coordinator.adjust_retrieval_threshold(0.75, 0.0);
        assert!((adjusted - 0.525).abs() < 0.01); // 0.75 * 0.7

        // High quality → higher threshold (can be selective)
        let adjusted = coordinator.adjust_retrieval_threshold(0.75, 1.0);
        assert!((adjusted - 0.75).abs() < 0.01); // 0.75 * 1.0

        // Medium quality
        let adjusted = coordinator.adjust_retrieval_threshold(0.75, 0.5);
        assert!((adjusted - 0.6375).abs() < 0.01); // 0.75 * 0.85
    }

    fn create_test_coordinator() -> SealKnowledgeCoordinator {
        let bks_cache = Arc::new(Mutex::new(
            BehavioralKnowledgeCache::in_memory(100).unwrap(),
        ));
        let pks_cache = Arc::new(Mutex::new(PersonalKnowledgeCache::in_memory(100).unwrap()));

        SealKnowledgeCoordinator::new(bks_cache, pks_cache, IntegrationConfig::default()).unwrap()
    }
}
